frontera/contrib/backends/redis_backend/__init__.py

```diff
@@ -0,0 +1,387 @@
+
+        results = []
+        for i in range(max_host_items):
+            for host_crc32, items in queue.items():
+                if len(items) <= i:
+                    continue
+                item = items[i]
+                (_, _, _, encoded, score) = item
+                to_remove.append(packb(item))
+                request = self._decoder.decode_request(encoded)
+                request.meta[FIELD_SCORE] = score
+                results.append(request)
+        if len(to_remove) > 0:
+            connection.zrem(partition_id, *to_remove)
+        return results
+
+    def schedule(self, batch):
+        to_schedule = dict()
+        now = int(time())
+        for fprint, score, request, schedule in batch:
+            if schedule:
+                # TODO: This is done by DomainMiddleware - RedisBackend should require DomainMiddleware
+                if FIELD_DOMAIN not in request.meta:
+                    _, hostname, _, _, _, _ = parse_domain_from_url_fast(request.url)
+                    if not hostname:
+                        self._logger.error("Can't get hostname for URL %s, fingerprint %s", request.url, fprint)
+                    request.meta[FIELD_DOMAIN] = {'name': hostname}
+                timestamp = request.meta[FIELD_CRAWL_AT] if FIELD_CRAWL_AT in request.meta else now
+                to_schedule.setdefault(timestamp, []).append((request, score))
+        for timestamp, batch in to_schedule.items():
+            self._schedule(batch, timestamp)
+
+    @classmethod
+    def get_interval_start(cls, score):
+        if score < cls.MIN_SCORE or score > cls.MAX_SCORE:
+            raise OverflowError
+        i = int(score / cls.SCORE_STEP)
+        if i % 10 == 0 and i > 0:
+            i -= 1  # last interval is inclusive from right
+        return i * cls.SCORE_STEP
+
+    def _schedule(self, batch, timestamp):
+        data = dict()
+        for request, score in batch:
+            domain = request.meta[FIELD_DOMAIN]
+            fingerprint = request.meta[FIELD_FINGERPRINT]
+            if type(domain) == dict:
+                partition_id = self._partitioner.partition(domain[FIELD_NAME], self._partitions)
+                host_crc32 = get_crc32(domain[FIELD_NAME])
+            elif type(domain) == int:
+                partition_id = self._partitioner.partition_by_hash(domain, self._partitions)
+                host_crc32 = domain
+            else:
+                raise TypeError("domain of unknown type.")
+            item = (timestamp, fingerprint, host_crc32, self._encoder.encode_request(request), score)
+            interval_start = self.get_interval_start(score)
+            data.setdefault(partition_id, []).extend([int(interval_start * 100), packb(item)])
+        connection = StrictRedis(connection_pool=self._pool)
+        pipe = connection.pipeline()
+        for key, items in data.items():
+            connection.zadd(key, *items)
```

Comment:
Then I would like to ask you to process them on every write attempt and take action: a) log error message explaining what happened, b) retry with some increasing interval till certain limit (say 5, 10, 30, 30, 30, 30... seconds).

I'm expecting that would be the best what we can do, e.g. halting the processing of spider-log or scoring-log until Redis will be ready to get new writes. It appears better to me than stopping.